plugin/pkg/admission/certificates/establishtrust/admission.go

```diff
@@ -0,0 +1,118 @@
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"k8s.io/apiserver/pkg/admission"
+	genericadmissioninit "k8s.io/apiserver/pkg/admission/initializer"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/component-base/featuregate"
+	"k8s.io/klog/v2"
+	api "k8s.io/kubernetes/pkg/apis/certificates"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/plugin/pkg/admission/certificates"
+)
+
+const PluginName = "ClusterTrustBundleEstablishTrust"
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return NewPlugin(), nil
+	})
+}
+
+// Plugin is the ClusterTrustBundle establishtrust plugin.
+//
+// In order to create or update a ClusterTrustBundle that sets signerName,
+// you must have the following permission: group=certificates.k8s.io
+// resource=signers resourceName=<the signer name> verb=establishtrust.
+type Plugin struct {
+	*admission.Handler
+	authz authorizer.Authorizer
+
+	inspectedFeatureGates bool
+	enabled               bool
+}
+
+var _ admission.ValidationInterface = &Plugin{}
+var _ admission.InitializationValidator = &Plugin{}
+var _ genericadmissioninit.WantsAuthorizer = &Plugin{}
+var _ genericadmissioninit.WantsFeatures = &Plugin{}
+
+func NewPlugin() *Plugin {
+	return &Plugin{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}
+}
+
+// SetAuthorizer sets the plugin's authorizer.
+func (p *Plugin) SetAuthorizer(authz authorizer.Authorizer) {
+	p.authz = authz
+}
+
+// InspectFeatureGates implements WantsFeatures.
+func (p *Plugin) InspectFeatureGates(featureGates featuregate.FeatureGate) {
+	p.enabled = featureGates.Enabled(features.ClusterTrustBundle)
+	p.inspectedFeatureGates = true
+}
+
+// ValidateInitialization checks that the plugin was initialized correctly.
+func (p *Plugin) ValidateInitialization() error {
+	if p.authz == nil {
+		return fmt.Errorf("%s requires an authorizer", PluginName)
+	}
+	if !p.inspectedFeatureGates {
+		return fmt.Errorf("%s did not see feature gates", PluginName)
+	}
+	return nil
+}
+
+var clusterTrustBundleGroupResource = api.Resource("clustertrustbundles")
+
+func (p *Plugin) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	if !p.enabled {
+		return nil
+	}
+	if a.GetResource().GroupResource() != clusterTrustBundleGroupResource {
+		return nil
+	}
+
+	newBundle, ok := a.GetObject().(*api.ClusterTrustBundle)
+	if !ok {
+		return admission.NewForbidden(a, fmt.Errorf("expected type ClusterTrustBundle, got: %T", a.GetOldObject()))
+	}
+
+	// Unlike CSRs, it's OK to validate against the *new* object, because
+	// updates to signer name will be rejected during validation.
+
+	// If signer name isn't specified, we don't need to perform the
+	// establishtrust check.
+	if newBundle.Spec.SignerName == "" {
+		return nil
+	}
+
+	if !certificates.IsAuthorizedForSignerName(ctx, p.authz, a.GetUserInfo(), "establishtrust", newBundle.Spec.SignerName) {
```

Comment:
I was talking with @enj to see if there was a less awkward verb we could use here... what would you think of `attest`? "affirm to be true or genuine", which is what trust bundles do for certificates, so anyone with write access is doing that for stuff issued by that signer.

Has the benefit of being an actual verb, and being used in security circles

Reply:
@mikedanese ?  (I have no real opinion on this verb).

I'll go ahead and assume that Mike is also fine with it, and make the change.

Reply:
Done.